library/core/src/slice/iter/macros.rs

```diff
@@ -158,25 +182,24 @@ macro_rules! iterator {
                 // one of the most mono'd things in the library.
 
                 let ptr = self.ptr;
-                let end_or_len = self.end_or_len;
                 // SAFETY: See inner comments. (For some reason having multiple
                 // block breaks inlining this -- if you can fix that please do!)
                 unsafe {
                     if T::IS_ZST {
-                        let len = end_or_len.addr();
+                        let len = self.end_addr_or_len;
                         if len == 0 {
                             return None;
                         }
                         // SAFETY: just checked that it's not zero, so subtracting one
                         // cannot wrap.  (Ideally this would be `checked_sub`, which
                         // does the same thing internally, but as of 2025-02 that
                         // doesn't optimize quite as small in MIR.)
-                        self.end_or_len = without_provenance_mut(len.unchecked_sub(1));
+                        self.end_addr_or_len = intrinsics::unchecked_sub(len, 1);
                     } else {
-                        // SAFETY: by type invariant, the `end_or_len` field is always
+                        // SAFETY: by type invariant, the `end_addr_or_len` field is always
                         // non-null for a non-ZST pointee.  (This transmute ensures we
                         // get `!nonnull` metadata on the load of the field.)
-                        if ptr == crate::intrinsics::transmute::<$ptr, NonNull<T>>(end_or_len) {
+                        if ptr == *(&raw const self.end_addr_or_len).cast::<NonNull<T>>() {
```

Comment:
Is this actually a good way to do this? You've been preferring using intrinsics directly due to optimisation concerns, and I worry this is running into the same problem.

It feels like something along the lines of `addr_usize(ptr) == self.end_addr_or_len` might be better.